### ISSUE TITLE
AttributeError when running in the context of an HTTP request

### ISSUE THREAD

**Issue by alewisohn** [bear/parsedatetime#110]
I get random occurrences of the following error:
```
AttributeError: 'thread._local' object has no attribute 'stack'
```
I'm using the library in the process of an HTTP request and I think that might be interfering with the use of thread locals.

**Comment by bear** [comment]
Which version of the library and Python are you using?

Also, what library are you using for the HTTP results -- Requests? or ???

**Comment by alewisohn** [comment]
Python 2.7.6
parsedatetime HEAD (version 1.6?)
Django REST Framework 2.4.2

I receive a POST from Twilio containing a text message that I then run through the `evalRanges` method before returning my response.

**Comment by bear** [comment]
You may be picking up some very experimental code that just landed - I would suggest trying the tagged v1.5 release at

https://github.com/bear/parsedatetime/tree/v1.5

until we have a chance to test and debug this code -- which your issue is the first!

**Comment by philiptzou** [comment]
Andrew:

Can you post the traceback information here so we can take a deep look
of this problem? Thank you.

Mike Taylor <notifications@github.com>于2015年7月30日星期四写道：

> You may be picking up some very experimental code that just landed - I
> would suggest trying the tagged v1.5 release at
>
> https://github.com/bear/parsedatetime/tree/v1.5
>
> until we have a chance to test and debug this code -- which your issue is
> the first!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bear/parsedatetime/issues/110#issuecomment-126097423>.
>


-- 
发自移动版 Gmail

**Comment by alewisohn** [comment]
```python-traceback
ERROR 2015-07-30 14:21:20,390 base 16657 140359787828992 Internal Server Error: ...
Traceback (most recent call last):
  File "/opt/envs/sitterfied/local/lib/python2.7/site-packages/django/core/handlers/base.py", line 111, in get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/opt/envs/sitterfied/local/lib/python2.7/site-packages/django/views/generic/base.py", line 69, in view
    return self.dispatch(request, *args, **kwargs)
  File "/opt/envs/sitterfied/local/lib/python2.7/site-packages/django/views/decorators/csrf.py", line 57, in wrapped_view
    return view_func(*args, **kwargs)
  File "/opt/envs/sitterfied/local/lib/python2.7/site-packages/rest_framework/views.py", line 401, in dispatch
    response = self.handle_exception(exc)
  File "/opt/envs/sitterfied/local/lib/python2.7/site-packages/rest_framework/views.py", line 398, in dispatch
    response = handler(request, *args, **kwargs)
  File "/opt/envs/sitterfied/local/lib/python2.7/site-packages/rest_framework/decorators.py", line 50, in handler
    return func(*args, **kwargs)
  File "/opt/sitterfied/sitterfied/flows/views.py", line 50, in booking_requests
    start, end, type = calendar.evalRanges(text, get_today().timetuple())
  File "/opt/envs/sitterfied/local/lib/python2.7/site-packages/parsedatetime/__init__.py", line 603, in evalRanges
    VERSION_CONTEXT_STYLE)
  File "/opt/envs/sitterfied/local/lib/python2.7/site-packages/parsedatetime/__init__.py", line 1817, in parse
    with self.context() as ctx:
  File "/usr/lib/python2.7/contextlib.py", line 17, in __enter__
    return self.gen.next()
  File "/opt/envs/sitterfied/local/lib/python2.7/site-packages/parsedatetime/__init__.py", line 293, in context
    self._ctxStack.push(ctx)
  File "/opt/envs/sitterfied/local/lib/python2.7/site-packages/parsedatetime/context.py", line 24, in push
    self.__local.stack.append(ctx)
AttributeError: 'thread._local' object has no attribute 'stack'
```

**Comment by philiptzou** [comment]
@alewisohn have a try? I think https://github.com/bear/parsedatetime/commit/be602b58b69f38d7feb0772e5ba497b82a6aa61b fixed this issue. It works for me.

**Comment by alewisohn** [comment]
:+1:
----
What's the goal behind the code changes? Are you just optimizing or are you adding new features?

**Comment by philiptzou** [comment]
Related pull-request here: https://github.com/bear/parsedatetime/pull/106

TL;DR: use new `Context` object to handle context values during parse and returned as the second value by `parse()` in new style

**Comment by alewisohn** [comment]
Cool, thanks.

**Comment by bear** [comment]
yes, this is very cool stuff IMO - thanks @philiptzou for fixing this!
